### DISCUSSION TITLE
Open Graph images not working anymore after updating to Next 13, despite being present in the HTML

### DISCUSSION THREAD

**Discussion by gmotta-dev** [vercel/next.js#50546] [Help]
### Summary

Since upgrading to the latest 13 version, I have encountered an issue where my Open Graph images are no longer appearing when I share my page. Interestingly, the images are present in the HTML code. I am using the metadata object inside the layout.ts

![image](https://github.com/vercel/next.js/assets/83229773/bff84971-13a9-49b0-adfb-94ecdb39b4f4)

Here is the link to the page: https://varos.com.br/vza

**Comment by icyJoseph** [comment]
Hi,

I do believe you are using `window` or `document`, as part of rendering, and that's crashing out.

Most precisely, your page is crashing when rendering on the server. How do I know this? Well I looked at the HTML sent by the server, and it is errored HTML.

Your app is then hydrated client side, and it looks like it works. Naturally bots miss out on this.

So, tell me, where do you think the app is crashing on the server? Are you perhaps using `"use client"`, expecting it to only render on the client?

**Reply by icyJoseph** [reply]
This what the server sends:

```html
<!DOCTYPE html>
<html id="__next_error__">
  <head>
    <meta charSet="utf-8" />
    <meta name="next-size-adjust" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <script src="[/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js](https://varos.com.br/_next/static/chunks/polyfills-c67a75d1b6f99dc8.js)" noModule=""></script>
  </head>
  <body>
</html>
```

I omitted the scripts for brevity.

**Reply by gmotta-dev** [reply]
I think I found out why, the problem was being caused by a font that I was importing from Adobe using the HTML Link <link rel="stylesheet" href="https://use.typekit.net/ail4fxb.css" />, when I remove this tag, apparently works fine. Do you think it could be it?

**Reply by icyJoseph** [reply]
Might be. You can check locally by doing `next build` and `next start`, then inspect the pages source code (left click view source in Chrome).

Not sure why the font would crash the SSR though.

**Reply by gmotta-dev** [reply]
I found out where the error is, but I dont know how I could solve this. The error is being caused if I use the useSearchParams Hook. If I use this hook and the metadata object at the same time, the OG Image is not displayed. Even if I use this hook inside a client component, it will break my OG Image. 

```
import { NextPage } from "next";

export const metadata = {
  title: "Parabéns pelo seu Valuation Box! – Valuation Box – Varos",
  description:
    'Parabéns pela sua assinatura! Sua assinatura do Valuation Box foi ativada com sucesso. Cheque sua caixa de entrada pelo e-mail com o assunto "Plataforma Varos - Acesso Liberado" para acessar a plataforma. Você também receberá um e-mail nosso comprovando a sua compra. Acessar Plataforma Varos © 2021.',
  openGraph: {
    images: [
      {
        url: "https://nextjs.org/og.png",
        width: 800,
        height: 600,
      },
    ],
  },
};

const MyPage: NextPage = ({}) => {
  return (
    <div>
      This is my page
      <MyButton />
    </div>
  );
};

export default MyPage;
```

```
'use client'

import {  useSearchParams } from "next/navigation";

const MyButton = ({}) => {

const params = useSearchParams()

  return (
    <button>
      Hello
    </button>
  );
};

export default MyButton;
```

**Reply by icyJoseph** [reply]
One kind of unrelated think, but you NextPage as a page type, I think, is gone.

I went to view-source:https://varos.com.br/vza and I see meta tags now, also I tested with the LinkedIn Post Inspector, and it looks like it can crawl your site and generate the OG data. 

https://www.linkedin.com/post-inspector/inspect/https:%2F%2Fvaros.com.br%2Fvza


So I guess, you fixed it?

**Comment by GustavoContreiras** [comment]
It's happening with me too. I can see things set on the final HTML but it's not working when I share link on Whatsapp.

**Reply by mxfng** [reply]
Encountered the same issue in my own app. useSearchParams appears to be the culprit for me too.

**Comment by alexhollender** [comment]
I'm having a similar issue. I'm currently using `next 14.0.3`. One of my sites uses `useSearchParams`, though removing it didn't seem to help. However another one of my sites _doesn't_ use `useSearchParams`, and is also having the same issue. As far as I can tell from the [metadata > opengraph-image documentation ](https://nextjs.org/docs/app/api-reference/file-conventions/metadata/opengraph-image) the only think I should need to do is place an `opengraph-image.jpg` file at the route/segment (which is what has worked for me in the past). However, since that's not working I tried explicitly defining the images in my `metadata` object:

```
import ogImage from './opengraph-image.jpg'
...
export const metadata = {
  ...
  openGraph: {
    ...
    images: [
      {
        url: ogImage.src,
        width: ogImage.width,
        height: ogImage.height
      },
    ],
  },
  twitter: {
    images: [
      {
        url: ogImage.src,
        width: ogImage.width,
        height: ogImage.height
      },
    ]
  },
}
```

This seems to fix the issue for now (though again, according to the documentation, doesn't seem like it should be required in order for this to work).

**Reply by TudorEsan** [reply]
not working for me, when i am trying to access the image it wants me to log in to vercel... any help?

**Reply by agusmoles** [reply]
Same here, it asks me for authentication for one project but not for the other one. Im using static metadata and `opengraph-image.png` file convention approach in both, but still working differently. 

EDIT: I fixed it using a hardcoded metadataBase instead of the autogenerated VERCEL_URL, which uses the deploy link instead of my own domain (which anyways should be public, but idk)

From:
```js
metadataBase: process.env.VERCEL_URL
    ? new URL(`https://${process.env.VERCEL_URL}`)
    : new URL(`http://localhost:${process.env.PORT || 3000}`),
```
to
```js
metadataBase: process.env.VERCEL_URL
    ? new URL(`https://agustinmoles.com.ar`)
    : new URL(`http://localhost:${process.env.PORT || 3000}`),
```

EDIT2:

Ok, this was the issue, I had it enabled by default for the new project this config:

![image](https://github.com/vercel/next.js/assets/28370654/00b15546-3e89-45c9-a6bb-e9004969b6b4)
which asks for auth for every domain except production ones and it seems somehow Vercel doesnt take autogenerated URL as a production link even though it is.

**Reply by Abdellah-Idrissi** [reply]
the solution 2 solved the problem fr me as well , but is it safe to nit anable the vercel auth ?

**Reply by anvodev** [reply]
The reason why NextJS don't appear og image properly is due to the lack of `metadataBase` (https://nextjs.org/docs/app/api-reference/functions/generate-metadata#metadatabase). I add metadataBase and it work for my site. Don't need to disable the Vercel Authentication.
Previous tag:
```
<meta name="twitter:image" content="https://anvodev-lc3fnlc4e-votanleans-projects.vercel.app/_next/static/media/token-bucket-cover.c9aa9bce.png">
```
After adding `metadataBase`:
```
<meta property="og:image" content="https://www.anvo.dev/_next/static/media/token-bucket-cover.c9aa9bce.png">
```

**Reply by menteshynov** [reply]
What works for me is to use `VERCEL_PROJECT_PRODUCTION_URL` instead of `VERCEL_URL`, this is also recommended in the Vercel documentation here [https://vercel.com/docs/projects/environment-variables/system-environment-variables](https://vercel.com/docs/projects/environment-variables/system-environment-variables)

**Reply by pearcircuitmike** [reply]
This is bizarre behavior and I don't feel like this *should* be the solution?

**Reply by pearcircuitmike** [reply]
> the solution 2 solved the problem fr me as well , but is it safe to nit anable the vercel auth ?

Is this the correct answer in general? I have the exact same problem but it seems weird that this would be the right move? Curious about the security implications too

**Comment by TudorEsan** [comment]
Hi for me it was not having: metadataBase: new URL("https://baseUrl") in my metadata, found this on stack overflow: https://stackoverflow.com/questions/76569441/next-js-13-og-image-not-working-when-i-put-it-in-app-directory

**Comment by DeepakBalaraman** [comment]
Turns out, the file size of the og:image could be a problem. 
For example, WhatsApp imposes a limit of 300 KB to opengraph images.

You can use tools like [image smaller](https://www.imagesmaller.com/) to compress images without losing quality.

For dynamically generated opengraph-image, try using image compression libraries like [sharp](https://github.com/lovell/sharp)

**Comment by ngquhuanbl** [comment]
Hello, after digging for a while, i found the root cause!
Disabling Vercel Authentication is a misleading solution, and doesn't address the root cause.
When Vercel deploys, the deployment is associated with a commit.
And the path you see of the og image is the URL of the deployment upon this commit.

Your production domain is just a custom domain, pointing to the result of the deployment of this commit, which is originally built targeting a commit-based domain.

When configuring the OpenGraph, for instance using the method mentioned in this issue report, I believe the author defining the url path for the OpenGraph images using relative URL. According to this [section](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#metadatabase) of  NextJS document, if the `metadataBase` is not provided, in production environment, `VERCEL_URL` will be used as a fallback. And [`VERCEL_URL`](https://vercel.com/docs/environment-variables/system-environment-variables#VERCEL_URL) is actually a System Environment Variable of Vercel, which mean it's automatically populated during the deployment. 
Therefore, when we deploy, Vercel will create a deployment for the commit, which has the `VERCEL_URL` corresponding to that commit. And the result of the `metadataBase` is the URL we see, with some commit has within it.

To resolve this, the official way is that we can created another environment variable, (which should not HAVE THE SAME name with any Vercel system environment variables), let's called it `PROD_URL`. Then, define `metadataBase` correctly as below:
```ts
// layout.tsx

// ...
import type { Metadata } from "next";

const defaultUrl = process.env.PROD_URL
  ? `https://${process.env.PROD_URL}`
  : "http://localhost:3000";

export const metadata: Metadata = {
  metadataBase: new URL(defaultUrl),
  // your other metadata (e.g. OpenGraph, etc.) go here
 }
```
Then, in your Vercel Environment settings, add a environment variables named `PROD_URL`, whose value is your production domain. Make sure this variables is visible in the Production environment. See instructions [here](https://vercel.com/docs/environment-variables/managing-environment-variables)

Here are a few other important notes:
- If you define the OpenGraph using the file-based approach (i.e. created the file following the conventional name rule in `/app`), at the same times, you define the configuration for the metadata using JS/TS code in `layout.tsx`. The configuration using JS/TS code won't be applied. I guess that the file-based approach has higher priority.
- Therefore, in order for the `metadataBase` to work, make sure you dont apply any file-based OpenGraph approach in your repo.
- You can put your images in the `public` folder, and define relative path to them in the metadata config. E.g. my image is at `public/shared/og-img.jpg`, then the relative path I use in my metadata config is `/shared/og-img.jpq`.

Hope this help!
